server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -6678,6 +6678,12 @@ public UserVm restoreVirtualMachine(final Account caller, final long vmId, final
                 newVol = volumeMgr.allocateDuplicateVolume(root, null);
             }
 
+            if (userVmDetailsDao.findDetail(vm.getId(), VmDetailConstants.ROOT_DISK_SIZE) == null && !newVol.getSize().equals(template.getSize())) {
+                VolumeVO resizedVolume = (VolumeVO) newVol;
+                resizedVolume.setSize(template.getSize());
+                _volsDao.update(resizedVolume.getId(), resizedVolume);
+            }
+
             // 1. Save usage event and update resource count for user vm volumes
             _resourceLimitMgr.incrementResourceCount(newVol.getAccountId(), ResourceType.volume, newVol.isDisplay());
             _resourceLimitMgr.incrementResourceCount(newVol.getAccountId(), ResourceType.primary_storage, newVol.isDisplay(), new Long(newVol.getSize()));
```

Comment:
`newVol` here doesn't have updated template size, refresh it after the template size is updated in DB.

Reply:
@sureshanaparti same object is been cast to VolumeVO and size is updated  at 6682-6683. `newVol` will return the updated size